DocumentStores/CosmosDb/Halforbit.DataStores.DocumentStores.CosmosDb/Implementation/CosmosDbDataStore.cs

```diff
@@ -610,7 +684,136 @@ async Task ObserveBeforePut(TKey key, TValue value)
 
             return tasks.SelectMany(t => t.Result);
         }
+        
+        private static async Task<TOut> OperationWrapper<TIn, TOut>(
+            SemaphoreSlim semaphore,
+            TIn input,
+            Func<TIn, Task<TOut>> asyncMutator)
+        {
+            await semaphore.WaitAsync().ConfigureAwait(false);
+            try
+            {
+                return await asyncMutator(input).ConfigureAwait(false);
+            }
+            finally
+            {
+                semaphore.Release();
+            }
+        }
+        
+        private async Task<IReadOnlyList<KeyValuePair<TKey, TOut>>> BulkOperation<TOut>(
+            IEnumerable<KeyValuePair<TKey, TValue>> values,
+            Func<Container, string, string, TKey, TValue, Task<TOut>> operation)
+        {
+            var partitionGroups = GroupByPartition(values);
+            var semaphore = new SemaphoreSlim(DataStoresConcurrency.MaxOperations);
+            var tasks = new List<Task<KeyValuePair<TKey,TOut>>>();
+            var bulkTasks = new List<Task<KeyValuePair<TKey, TOut>[]>>();
+            
+            foreach (var grp in partitionGroups)
+            {
+                foreach (var batch in grp.Batch(BulkApiMaxOpsPerRequest))
+                {
+                    var useBulk = batch.Count() > BulkApiMinOpsPerPartition;
+                    if (useBulk)
+                    {
+                        bulkTasks.Add(OperationWrapper(semaphore, batch,
+                            async b =>
+                            {
+                                var results = b.Select(async i => new KeyValuePair<TKey, TOut>(i.Key,
+                                    await operation(_bulkContainer.Value, i.PartitionKey, i.DocumentId, i.Key, i.Value)));
```

Comment:
we probably need to put a `.ConfigureAwait(false)` on _all_ usages of `await`, both here and elsewhere in these changes.

Reply:
I tried to be cognizant of that, apparently I missed a bunch on the changes in this file.

Reply:
Seemed like a large scale problem throughout the code base. So I went ahead and addressed them all.

5e11e77d6c22b1bc6bfd8dcd266f86e06476bfd7